pinot-core/src/main/java/org/apache/pinot/core/operator/combine/SortedGroupByCombineOperator.java

```diff
@@ -0,0 +1,250 @@
+  private volatile DataSchema _dataSchema;
+
+  private final AtomicReference<SortedRecords> _waitingRecords;
+  private final Comparator<Record> _recordKeyComparator;
+  private final SortedRecordsMerger _sortedRecordsMerger;
+
+  public SortedGroupByCombineOperator(List<Operator> operators, QueryContext queryContext,
+      ExecutorService executorService) {
+    super(null, operators, overrideMaxExecutionThreads(queryContext, operators.size()), executorService);
+
+    assert (queryContext.shouldSortAggregateUnderSafeTrim());
+    _operatorLatch = new CountDownLatch(_numTasks);
+    _waitingRecords = new AtomicReference<>();
+    _recordKeyComparator = OrderByComparatorFactory.getRecordKeyComparator(queryContext.getOrderByExpressions(),
+        queryContext.getGroupByExpressions(), queryContext.isNullHandlingEnabled());
+    _sortedRecordsMerger =
+        GroupByUtils.getSortedReduceMerger(queryContext, queryContext.getLimit(), _recordKeyComparator);
+  }
+
+  /**
+   * For group-by queries, when maxExecutionThreads is not explicitly configured, override it to create as many tasks as
+   * the default number of query worker threads (or the number of operators / segments if that's lower).
+   */
+  private static QueryContext overrideMaxExecutionThreads(QueryContext queryContext, int numOperators) {
+    int maxExecutionThreads = queryContext.getMaxExecutionThreads();
+    if (maxExecutionThreads <= 0) {
+      queryContext.setMaxExecutionThreads(Math.min(numOperators, ResourceManager.DEFAULT_QUERY_WORKER_THREADS));
+    }
+    return queryContext;
+  }
+
+  @Override
+  public String toExplainString() {
+    return EXPLAIN_NAME;
+  }
+
+  /**
+   * Executes query on one sorted segment in a worker thread and merges the results into the sorted record table.
+   */
+  @Override
+  protected void processSegments() {
+    int operatorId;
+    while (_processingException.get() == null && (operatorId = _nextOperatorId.getAndIncrement()) < _numOperators) {
+      Operator operator = _operators.get(operatorId);
+      try {
+        if (operator instanceof AcquireReleaseColumnsSegmentOperator) {
+          ((AcquireReleaseColumnsSegmentOperator) operator).acquire();
+        }
+        GroupByResultsBlock resultsBlock = (GroupByResultsBlock) operator.nextBlock();
+        if (resultsBlock.isGroupsTrimmed()) {
+          _groupsTrimmed = true;
+        }
+        // Set groups limit reached flag.
+        if (resultsBlock.isNumGroupsLimitReached()) {
+          _numGroupsLimitReached = true;
+        }
+        if (resultsBlock.isNumGroupsWarningLimitReached()) {
+          _numGroupsWarningLimitReached = true;
+        }
+        if (_dataSchema == null) {
+          _dataSchema = resultsBlock.getDataSchema();
+        }
+        // short-circuit one segment case
+        if (_numOperators == 1) {
+          _waitingRecords.set(GroupByUtils.getAndPopulateSortedRecords(resultsBlock));
+          break;
+        }
+
+        SortedRecords records =
```

Comment:
This involves extra memory overhead. We can directly put `resultsBlock` in, and when getting the reference out, do a `instanceof` check and only extract the records when needed